[1 - 2단계 방탈출 예약 대기] 드라고(이용화) 미션 제출합니다.

Firedrago95 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
 [방탈출 대기 1단계 프롤로그](https://prolog.techcourse.co.kr/studylogs/4321)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
[이번 미션에서 변경된 부분🔗](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/files/f388a7743331b9ad42dff62d90e3e674a7f82252..0412ed57b460b4b0ad48c71717e0206a48bf369f)
안녕하세요 제이미~ 이번 미션 리뷰 잘 부탁드립니다.

### 1. Jpa와 불변객체
- JPA에서는 엔티티 필드에 final을 쓸 수 없다는 제약이 있어서, 이를 해결하기 위해 원시값을 포장하고 @Embedded로 불변성을 보장하자는 의견이 있었습니다. 이 방식이 객체의 불변성과 설계 측면에서 장점이 있는 건 이해되지만, JPA와의 호환성이나 퍼포먼스 측면에서 문제가 발생할 여지는 없을까요?
### 2. Jpa 개념의 적절한 사용
- Jpa 를 적용하는데에 있어 적절한 기능을 사용하였는지 궁금합니다.
### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kkiseug committed:
feat: 프로젝트 구성

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
chore: Spring Data Jpa 의존성 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
chore: Jpa 관련 설정 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 기존 리포지토리 구현체 Spring Data jpa 구현체로 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 테스트 스프링 설정 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 운영 초기 데이터 sql 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: ThemeService 테스트 config 주입으로 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
fix: 테마 id 조회 메서드 네이밍 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 비즈니스 예외 처리 400 http 예외 코드로 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 엔티티 멤버 속 VO record -> class

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 테스트 config 주입으로 변경

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 프론트 코드 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 레거시 코드 삭제

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 프론트 코드 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 예약 상태 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 예약 상태 추가로 인한 sql문 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 멤버와 예약 양방향 매핑

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 멤버별 예약 조회 기능 구현

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
test: 멤버별 예약 조회 테스트 작성

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
chore: 테스트 스키마 제거

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 멤버별 예약 조회 책임 Reservation 도메인으로 이동

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
refactor: 예약 api 컨트롤러 역할별 분리

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
style: 컨벤션 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
feat: 멤버 예외 추가

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

kkiseug committed:
fix: 컨트롤러 예외 처리 수정

Co-authored-by: Firedrago95 <yonghwa95@naver.com>

jamie9504 commented on build.gradle:
```diff
@@ -20,7 +20,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter'
```
> [이전 미션 pr 🔗](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/files/f388a7743331b9ad42dff62d90e3e674a7f82252..6206cfd0319bd73a2f32aacc0e5a481e8dcfe378)

이게 무슨 의미인지 유추할 수 있나요?

---

README를 정리하지 않은 이유가 있나요?

Firedrago95 replied:
1. '이번 미션에서 변경된 부분' 이라는 문구로 변경했습니다.

2. README를 작성하였습니다.

[docs: readme 작성](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/68be597800f3c96fbe319498b62a42ce605b60fa)

jamie9504 replied:
> [이번 미션에서 변경된 부분🔗](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/files/f388a7743331b9ad42dff62d90e3e674a7f82252..6206cfd0319bd73a2f32aacc0e5a481e8dcfe378)

PR 본문은 PR이 열려있는 동안 지속적으로 관리되어야 합니다.
이번 미션에서 변경된 수정 사항을 모두 포함하고 있나요?

Firedrago95 replied:
추가된 커밋을 모두 포함하도록 수정하였습니다.

jamie9504 commented on src/main/java/roomescape/admin/infrastructure/JpaAdminRepositoryAdaptor.java:
```diff
@@ -0,0 +1,31 @@
+import roomescape.admin.domain.AdminRepository;
+
+@Repository
+public class JpaAdminRepositoryAdaptor implements AdminRepository {
```
아마도 Adaptor를 사용했던데, 드라고도 Adaptor를 사용했군요. 동일한 숙제를 내겠습니다.

1. Adaptor를 사용했을 때의 장/단점은 무엇이며, 현재 프로젝트에서 사용했을 때의 장/단점은 무엇인가요?

2. JDBC -> JPA 전환 관점과 향후 유지보수 관점에서 1을 보강해보세요.

3. [숙제] 해당 부분에 대한 페어 외의 크루 4명 이상과 1, 2에서 정리한 내용을 이야기해보고 이야기 후 알게 된 점이나 느낀 점에 대해 정리해주세요.

Firedrago95 replied:
Repository에 Adaptor를 사용하면 크게 4가지 장점이 있다고 생각합니다.
1. dip, isp 원칙을 지키며 db변경에 유연함을 가질 수 있습니다. jpa에서 다른 db연결 기술을 변경하는 경우에, 기존 코드의 변경없이 Adaptor가 의존하는 구현체만 변경하면 되는 유연한 구조를 만들 수 있습니다.
2. 도메인 레이어는 구체적인 JPA 구현체에 직접 의존하지 않기 때문에, 기존 JDBC 기반의 Repository 구현체를 그대로 유지하거나 병렬로 두고 테스트를 실행할 수 있습니다.
3. fake 객체를 사용하여 테스트를 작성 할 수 있습니다. 테스트에서는 JPA나 DB 없이도 필요한 기능을 검증할 수 있고, 도메인 로직은 데이터 저장 방식과 완전히 분리되어 있어 유지보수와 기술 교체에도 유리합니다.
4. 메서드에 적절한 이름을 부여할 수 있습니다. spring-data-jpa를 사용하는 경우, 메서드명의 규칙이 정해져 있기때문에, 메서드명이 길어지거나, 원하는 이름을 부여하기 어렵지만, adaptor로 감싸서 원하는 메서드 이름을 부여할 수 있습니다.

단점은 클래스 수가 증가하고 구조가 복잡해질 수 있습니다.

 토의 과정에서 느낀 점은, 다른 사람의 의견을 받아들이기 위해선 의식적인 노력이 필요하다는 것이었습니다. 처음에는 페어의 Adapter 제안에 거부감이 들었지만, 근거를 듣고 제이미의 질문에대한 고민을 이어가면서 오히려 내가 쓸 수 있는 도구가 하나 늘어났다는 점에서 배움이 되었습니다. 현재 프로젝트에서 jdbc 기존 코드를 유지하지도 않고, 테스트코드에서 fake 객체를 사용하지 않지만, 향후 변경가능성을 고려한다면, 가치가 있다고 느꼈습니다. 앞으로는 필요성만으로 배움을 제한하지 않고, 다양한 의견을 수용하면서 내가 쓸 수 있는 도구를 늘린다는 마음가짐으로 노력해야 한다는 것을 느꼈습니다.

jamie9504 replied:
토의 과정에서 좋은 점을 발견하셨네요.
그렇다면 반대로 Adaptor를 사용/학습하면서 잃은 점은 무엇인가요?

모든 것은 trade-off이므로 장점을 생각할 땐 꼭 단점도 함께 고민해보세요.

Firedrago95 replied:
Adapter를 사용하면서 느낀 가장 큰 단점은 복잡성 증가였습니다. 단순히 DB 연결을 위한 기능을 구현함에도 불구하고, 인터페이스, 구현체, 어댑터까지 총 3개의 클래스가 필요해졌고, 이로 인해 코드의 흐름을 빠르게 이해하기 어려워졌습니다.

또한 현재처럼 요구사항이 단순한 경우에는 오히려 과한 구조처럼 느껴질 수 있었고, 향후 기능이 추가될 경우 복잡성이 기하급수적으로 증가할 수 있다는 우려도 있었습니다. Adapter 패턴은 확장성과 유연성을 위한 좋은 도구이지만, 확장성이 요구되는 경우에 고려해야 한다는 생각을 했습니다.

jamie9504 replied:
JDBC에서 JPA로 사용 툴을 바꾸는 것이 이번 과제였지요.
밑의 코멘트와 약간 연결지어서, `(드라고가 이번 미션을 통해 얻고 싶은 부분이 무엇인지, 목표나 방향에 대해 얘기해주세요) 요 부분에 대한 코멘트를 먼저 작성 후 해당 코멘트에 대한 답변을 작성하기를 권장드려요.`
해당 장/단점이 드라고의 목표나 방향과 부합했다고 생각하시나요?

Firedrago95 replied:
'JPA 학습'을 기준으로 볼때 Adaptor 의 도입은 학습목표에 부합한다고 생각합니다. DB 연결 기술이 바뀌더라도, 기존 코드에 최소한의 수정만으로 대응할 수 있는 구조를 학습할 수 있었기 때문입니다. 물론 현재 프로젝트에서는 Adaptor가 제공하는 장점 중 일부(예: 테스트의 유연성)는 크게 드러나지 않았습니다. 하지만 도입 과정에서 장점과 단점을 고민하며 얻게 된 지식 덕분에, 향후 테스트를 더 유연하게 구성해야 하는 상황에서 유용한 도구로 고려할 수 있게 되었다고 생각합니다. 또한 복잡성과 유연함의 트레이드 오프를 고민해 볼 수 있었던 기회였기 때문에 JPA 기술 도입 과정에서 고민해봐야할 점들을 알 수 있었다고 생각합니다. 
 추가적으로 새로운 의견에 열린마음으로 장단점을 탐색해야 한다는 생각도 할 수 있었습니다.

jamie9504 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -1,13 +1,35 @@
-    private final Name name;
-    private final Email email;
-    private final Password password;
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
다음을 알아보고 간단히 정리해보세요.

1. ID를 설정하는 방식들
2. Enum을 저장할 수 있는 방식들
3. 그 외 이번에 처음 사용한 애너테이션들과 그 속성들

Firedrago95 replied:
ID를 설정하기 위한 방식은 IDENTITY, SEQUENCE, AUTO 방식이 있습니다. 
먼저 `GenerationType.IDENTITY`는 ID를 데이터베이스가 직접 생성하는 방식입니다. INSERT가 실행될 때 DB가 자동으로 ID를 부여합니다. 영속성 컨텍스트에 등록하기 위해 id가 필요하기 때문에 이 방식에서는 persist()가 호출되면 JPA는 즉시 INSERT 쿼리를 실행하여 DB가 생성한 ID 값을 가져옵니다.

반면 `GenerationType.SEQUENCE`는 NSERT를 실행하기 전에 DB 시퀀스 객체에서 미리 ID를 받아옵니다. 즉, persist() 시점에 시퀀스를 조회하고 ID를 설정한 후, 실제 INSERT는 flush() 시점에 발생합니다.

`GenerationType.AUTO`는 JPA가 사용하는 데이터베이스를 기준으로 IDENTITY나 SEQUENCE 전략을 자동으로 선택하는 방식입니다.

Firedrago95 replied:
Enum을 저장하기 위한 방식은 두가지가 있습니다.
`@Enumerated(EnumType.STRING)` 애너테이션을 사용하면 enum 타입의 값이 문자열의 형태로 db에 저장됩니다. 새로운 enum 이 추가되어도 문제없이 사용가능 합니다.
`@Enumerated(EnumType.ORDINAL)` 애너테이션을 사용하면 enum 타입의 인덱스 순서로 db에 저장됩니다. 중간에 새로운 enum 이 추가되면 기존의 데이터에 문제가 생길 수 있습니다.

jamie9504 replied:
> 그 외 이번에 처음 사용한 애너테이션들과 그 속성들

에 대한 정리도 보강해주세요.

---

> Enum을 저장할 수 있는 방식들

은 두 가지 방법이 다인가요?

Firedrago95 replied:
- 엔티티와 DB를 매핑하면서 `@Embedded`와 `@Embeddable` 애너테이션을 처음 사용해보았습니다. 엔티티의 필드를 객체 단위로 묶어서 표현하고 싶을 때, 값 객체에 `@Embeddable`을 선언하고, 엔티티에서는 해당 필드에 `@Embedded`를 사용하여 매핑할 수 있습니다. db 테이블은 분리되지 않지만, 객체의 응집도를 높이고 재사용성을 높일 수 있었습니다. 테이블 필드를 객체로 감싼 만큼 중복되는 이름이 생길 위험이 증가합니다. 이때는 ` @AttributeOverrides`를 반드시 사용해 이름을 재정의해야 합니다.

- @DataJpaTest는 JPA가 생성한 Repository와 관련된 Bean만 로드하며, 서비스나 컨트롤러와 같은 웹/비즈니스 레이어는 로드되지 않습니다. 기본적으로 인메모리 db를 자동으로 설정하고, 각 테스트는 트랜잭션으로 감싸져 실행되며 테스트가 끝나면 자동으로 롤백됩니다. 덕분에 테스트 간 데이터가 서로 영향을 주지 않아 독립적이고 반복 가능한 테스트가 가능합니다

Firedrago95 replied:
enum을 저장하는 새로운 방식이 무엇이 있는지 찾지 못하였습니다. 혹시 다른 방법이 있다면 알려주실수 있을까요?

jamie9504 replied:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/8fa65a08-8932-43f3-8d64-564ecf7dff8d" />

https://techblog.woowahan.com/2600/

새로운 방식을 찾기 위해 다양한 방법을 사용해보세요.
검색 키워드를 다양화해보는 것도 좋습니다.
AI에게 물어보던, 구글링을 하던 방식에 구애받지 않고 드라고가 원하는 답을 찾을 수 있는 다양한 방식을 만들어보세요.

Firedrago95 replied:
ai를 활용하여 enum 타입에 대한 Converter 방식이 있다는 것을 알게되었습니다. 적용해보았습니다.

[feat: StatusConverter 추가 및 Converter 적용](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/2b93a38927d0d3726dcd7f88c5eab9923effa5e2)

jamie9504 replied:
넓게 찾는 법과 정확히 찾는 법 둘 다 앞으로도 중요할 것이므로, 방법들을 많이 잘 알아둘수록 좋습니다.

jamie9504 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -1,13 +1,35 @@
+    @Embedded
+    private Password password;
+
+    @OneToMany(mappedBy = "member", fetch = FetchType.LAZY)
```
`@OneToMany` fetch의 기본값은 무엇인가요?
다른 전략들은 무엇이 있나요?

Firedrago95 replied:
`@OneToMany`의 fetch 기본값은 LAZY입니다. 즉, 연관된 엔티티는 실제로 접근할 때 조회됩니다. 만약 fetch = EAGER로 설정하면, 모든 Member를 조회할 때 각 Member가 가진 Reservation도 즉시 함께 조회되며, 이로 인해 N+1 문제를 유발할 수 있습니다.

이러한 이유로 JPA는 ~ToMany 관계에서는 기본적으로 LAZY를 사용하고, 반대로 `@ManyToOne`, `@OneToOne` 같은 ~ToOne 관계에서는 기본값이 EAGER입니다. ~ToOne 관계는 일반적으로 추가 조회가 한 건이기 때문에 성능 부담이 적어 기본적으로 즉시 로딩을 사용합니다.

jamie9504 replied:
기본 전략임에도 명시해준 이유가 있었나요?

Firedrago95 replied:
기본 전략이 모르는 사람이 보더라도, 지연로딩 된다는 것을 명시적으로 보여줄 수 있기 때문입니다. 지연로딩으로 설정하지 않았을 경우 N+1문제로 인해 문제가 발생할 수 있으므로 기본 전략이지만, 명시해주는것이 좋다고 생각했습니다.

jamie9504 replied:
기본 전략을 모르는 사람을 위해, `FetchType.LAZY`를 명시한다면 모든 기본 전략을 명시해주어야 일관성이 있는게 아닌가요?

> 지연로딩으로 설정하지 않았을 경우 N+1문제로 인해 문제가 발생할 수 있으므로 기본 전략이지만, 명시해주는것이 좋다고 생각했습니다.

`FetchType.LAZY`가 N+1 문제를 완전히 방지할 수 있나요?

Firedrago95 replied:
제가 잘못 알고 있었던 부분이 있었습니다. LAZY 전략과 EAGER 전략은 N+1과 관련있는 것이 아니라 조회 시점에 연관관계를 가진 엔티티를 함께 조회할지, 프록시 객체를 대신 사용하여 필요할때 조회하는지의 차이가 있었습니다. EAGER 전략은 연관된 엔티티를 항상 함께 로딩하기 때문에 개발자가 제어하기 어려워 예상치 못한 N+1 문제가 발생할 수 있습니다. 반면 LAZY 전략은 프록시 객체를 통해 연관 엔티티를 필요한 시점에 지연 로딩합니다. 다만, LAZY라고 해서 N+1 문제가 자동으로 해결되는 것은 아니고, 루프 내에서 연관 객체에 반복적으로 접근하면 똑같이 N+1 문제가 발생할 수 있습니다. N+1 문제를 해결하려면 LAZY 전략을 기반으로 하고, JPQL의 fetch join이나 `@EntityGraph` 등을 활용해 필요한 시점에 한 번의 쿼리로 연관 데이터를 미리 로딩하는 방식이 가장 안정적인 해결책이라는 점을 정리할 수 있었습니다.

일관성을 위해 기본 전략을 명시하지 않도록 변경했습니다.

[feat: Member 엔티티의 fetch 전략 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/55b32f43de3469d8dfe4d0b884baef73d5deba90)

jamie9504 replied:
좋아요. 처음 학습하는 부분은 헷갈릴 수 있어요.
JPA를 학습한다. 자체도 중요하지만 무언가를 학습할 때 어느 깊이로 어떤 방향으로 학습을 해야할지 이번 과정을 통해 체화해두면
다른 기술을 학습할 때 시행착오를 덜 겪을 수 있을거에요.

jamie9504 commented on src/main/java/roomescape/admin/service/AdminService.java:
```diff
@@ -14,14 +15,16 @@ public AdminService(AdminRepository adminRepository) {
     }
 
     public boolean isExistsByEmail(final String email) {
```
JPA를 적용했음에도 트랜잭션에 대한 고려가 없네요.
트랜잭션에 대한 적절한 학습 후 `@Transactional` 애너테이션에 대해 간단히 정리하고 사용해보세요.

Firedrago95 replied:
`@Transactional`을 사용하지 않을 경우, 하나의 서비스 메서드 내에서 각 repository 메서드 호출마다 별도의 트랜잭션이 시작되고 종료됩니다. 하지만 서비스 메서드에 `@Transactional`을 적용하면, 스프링이 해당 메서드에 대해 프록시 객체를 생성하고, 메서드 실행 전 트랜잭션을 시작하며, 실행 완료 후 트랜잭션을 커밋 또는 롤백합니다. 이로 인해 하나의 서비스 메서드 전체가 하나의 트랜잭션 범위 내에서 실행된다는 점을 알게되었습니다. 
 또한 읽기 작업만 수행하는 경우 readOnly 속성을 true 로 설정하여, jpa 내부의 변경감지 기능을 사용하지 않아서 성능을 더욱 최적화 할 수 있습니다.
 
 [refactor: Transactional 어노테이션 도입](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/788a488e4626a9f8bbc0a5393a6c67a7efafec6d)

jamie9504 commented on src/main/java/roomescape/theme/infrastructure/JpaThemeRepository.java:
```diff
@@ -0,0 +1,29 @@
+        ORDER BY count DESC
+        LIMIT :limit
+        """, nativeQuery = true)
+    List<Theme> findPopularThemes(@Param("start_date") LocalDate start, @Param("end_date") LocalDate end, @Param("limit") int popularCount);
```
nativeQuery 말고 해결할 방법은 없는걸까요?

---

CrudRepository를 쓴 이유는 무엇인가요?

Firedrago95 replied:
처음 미션을 수행할때, spring-data-jpa 의 생성 기능으로 표현하기 힘든 작업을 표현하는 방법을 페어와 논의했습니다.
1. nativeQuery를 그대로 사용하는 방법은 기존 jdbc에서 작성된 쿼리를 다시 재사용할 수 있어 편리하다고 생각했습니다.
2. JPQL jpa를 사용할때, jPQL이라는 SQL과 유사하면서도 상대적으로 객체적인 표현을 사용하는 표기방법이 있다는 것을 알고 있었습니다. 그런데 페어와 제가 둘 다 아직 익숙하지 않아서 도입하지 않았습니다.
3. QueryDsl 기술은 컴파일 단계에서 에러를 잡아낼 수 없는 SQL 문의 단점을 보완하기 위해 메서드 체이닝을 통해 SQL문을 표현할 수 있는 기술로 알고있습니다. 아직 익숙하지 않을 뿐더러 새로운 기술을 도입할 만큼 필요성이 크다고 생각하지 않아서 사용하지 않았습니다.

Firedrago95 replied:
`CrudRepository`를 사용하게된 이유는 사실 LMS의 예제코드를 그대로 도입했기 때문이었습니다. 
제이미가 리뷰를 남겨주신 이후에 `JpaRepository`와의 차이를 정리 했습니다. 두 Repository는 비슷한 기능을 수행하지만,
`JpaRepository`는 페이징과 정렬기능이 조금 더 추가되어 있다는 것을 알게되었습니다. `CrudRepository`를 사용하면 명시적으로 페이징과 정렬같은 부가적 기능을 사용하지 않고, CRUD 기능을 표현할 수 있다는 점을 명시하는 용도로 사용할 수 있다는 것을 배웠습니다.

jamie9504 replied:
> nativeQuery를 그대로 사용하는 방법은 기존 jdbc에서 작성된 쿼리를 다시 재사용할 수 있어 편리하다고 생각했습니다.

JDBC에서 사용한 방식을 그대로 사용한다면 JPA로 바꾼 의의가 있을까요?

> JPQL jpa를 사용할때, jPQL이라는 SQL과 유사하면서도 상대적으로 객체적인 표현을 사용하는 표기방법이 있다는 것을 알고 있었습니다. 그런데 페어와 제가 둘 다 아직 익숙하지 않아서 도입하지 않았습니다.

Adaptor를 도입할 때도, 익숙하지 않았을 텐데 도입했을텐데요.
드라고는 무언가를 도입할 때 우선순위가 무엇인가요?

Firedrago95 replied:
이번 미션에서는 QueryDSL이나 JPQL이 익숙하지 않은 상황에서, nativeQuery가 기능적으로 더 명확하고 빠르게 구현 가능하다고 판단했습니다. 물론 Adaptor 도입처럼, 처음엔 낯설지만 명확한 장점이 보일 때는 학습 비용을 감수하고 도입하는 방향을 택합니다. 그에 비해 이번 쿼리는 도메인 설계나 테스트 전략에 큰 영향을 주지 않는 범위라고 판단해, 우선순위를 낮췄습니다.
 또한 count() 함수와 limit 등은 JPQL 에서 지원되지 않는 기능이라는 것을 알게되었습니다. 또한 쿼리 하나를 표현하기위해 QueryDsl 을 도입하는 것 역시 비 효율적이라 판단했습니다.

jamie9504 replied:
위에 정의해준 부분은 `페어와 미션을 하는 기간동안`의 판단 기준이었다고 생각해요.
드라고가 이번 미션을 통해 얻고 싶은 부분이 무엇인지, 목표나 방향에 대해 얘기해주세요.

Firedrago95 replied:
저는 이번 미션을 통해 JPA의 사용과 적용에 필요한 지식을 습득하고 싶습니다. 그런데 미션 진행 과정을 돌아보면 새로운 기술을 도입할때의 우선순위나 기준 없이 그때그때 판단했다는 생각이듭니다. 앞으로 미션을 진행할때도 이번 미션에서 내가 배워야할 지식은 무엇인지 목표는 무엇인지 먼저 생각하고, 새로운 기술이 그 목표와 지식에 부합하는지 생각하는 습관을 만들어 보려고 합니다.

 'JPA 학습'을 기준으로 nativeQuery를 그대로 사용하는 것이 적절한지 생각해보았습니다. JPA는 데이터베이스와 객체지향의 패러다임 차이를 극복하기 위해 객체지향적인 방식으로 데이터베이스와 통신하기위해 등장했다고 생각합니다. JPA는 관계형 데이터베이스와 객체지향 간의 패러다임 차이를 극복하기 위해 등장한 기술로, 테이블 중심의 SQL이 아닌, 객체 중심의 명령 방식인 JPQL을 제공한다고 알고 있습니다. 그렇다면 JPA를 사용하는 이상, 객체지향적인 방식으로 데이터베이스와 통신하는 JPQL을 학습하고 적용해보는 것이 더 의미 있는 방향이라고 판단하게 되었습니다.
 
 [refactor: 인기 테마 조회 로직 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/8ffc4b4bfe9ca2d9691ccec4c6c99700d06f3792)

jamie9504 replied:
앞으로 미션은 하나, 단계는 3개밖에 남지 않았지만
무언가를 할 때 내가 하고자하는 목표가 무엇인지 세우고 그것을 다 충족한 후 +a를 하는 것과,
+a를 하다보니 목표와 벗어나서 다 충족하지 못하는 것은 같은 노력을 해도 결과적으로 차이가 있을 수 있겠죠?

모든걸 다 학습해야 할 순 있지만, 그 학습에서의 우선순위도 분명 있을거에요.
`이번 미션을 통해 JPA의 사용과 적용에 필요한 지식 습득을 한다.`는 큰 목표를 가졌다면,
세부 목표와 우선순위 정리를 추가적으로 하면서 남은 기간 어떻게 다음 단계를 구현하고 리팩터링할지 가이드라인을 잡아볼까요?

이 내용은 정리해서 다음 단계 PR 올릴 때 본문에 함께 적어주세요.

jamie9504 commented on src/main/java/roomescape/member/domain/Email.java:
```diff
@@ -1,15 +1,20 @@
 import java.util.Objects;
 import roomescape.member.exception.EmailException;
 
+@Embeddable
```
> JPA에서는 엔티티 필드에 final을 쓸 수 없다는 제약이 있어서, 이를 해결하기 위해 원시값을 포장하고 @Embedded로 불변성을 보장하자는 의견이 있었습니다. 이 방식이 객체의 불변성과 설계 측면에서 장점이 있는 건 이해되지만, JPA와의 호환성이나 퍼포먼스 측면에서 문제가 발생할 여지는 없을까요?

`@Embedded`이 JPA에서 제공하는 애너테이션인데요.
어떤 부분에서 문제가 발생할 것으로 추측 혹은 걱정하고 있었을까요? 드라고를 고민하게 한 포인트가 궁금해요.

---

별개로 `@Embedded`를 사용했을 때 있던 장/단점과 유의해야 할 점은 무엇이 있을까요?

Firedrago95 replied:
`@Embedded` 어노테이션을 사용하여 값 객체로 감싸게 되면, 원시값을 의미 있는 타입으로 감싸고, 관련 로직을 그 안에 함께 담을 수 있다는 점에서 코드의 명확성과 유지보수성이 높아진다고 느꼈습니다. 고민이 되었던 지점은 @Embedded가 JPA 내부 동작 방식과 충돌할 여지가 있지는 않을지 궁금했습니다. 엔티티의 경우에는 final 키워드를 필드에 사용할 수 없었기 때문에, record 값 객체로 정의하는 것이 문제가 없는지 궁금했습니다.
 또한 static 전역 상수 객체를 공유하게 되면 모든 엔티티 인스턴스가 같은 객체를 참조하게 되고, 이로 인해 영속성 컨텍스트 충돌이나 데이터 누락 같은 치명적인 버그가 발생할 수 있다는 점을 알게되었습니다. @Embedded는 응집도 높은 객체 모델링과 재사용성을 높이는 데 유용한 기능이지만, 불변성을 구현한다고 해서 전역 상수를 사용하는 것은 오히려 위험을 초래할 수 있고, JPA의 제약을 잘 이해한 상태에서 사용하는 것이 중요하다는 점을 깨달았습니다.

jamie9504 replied:
`@Embedded`를 JPA에서 제공하는 것은 맞으므로 이를 사용하는 것은 JPA 방식과 충돌하지 않을테지만
프레임워크에서 사용하도록 제공하는 방식과 다른 방식의 사용은 충분히 문제가 될 수 있겠죠.

무언가의 라이브러리, 프레임워크를 사용할 땐 어떠한 방식을 제공하고 있고 어떤 식으로 사용하도록 제공하고 있는지도 고민해봐야해요.

이와 별개로 불변을 구현해야만 하는 이유는 무엇인지도 고민해보면 좋아보이네요.

Firedrago95 replied:
`@Embedded` 를 이용하여 값객체로 감쌀 경우, 재사용성이 높아질 수 있습니다. 그렇다면 상대적으로 더 많은 곳에서 객체를 참조할 가능성이 높아지고, 객체의 변경가능성이 높아집니다. 이럴 경우 불변객체로 설계하여, 변경에 대해 신뢰를 주는 것이 이점이 있을 수 있다는 생각이 들었습니다.

jamie9504 replied:
좋아요. JPA에서 final을 사용할 수 없는 이유는 프록시 생성 및 지연 로딩 때문인데요. `@Embedded` 값 객체를 불변으로 만드는 것은 JPA를 사용할 때 문제가 없습니다.

Java Record 역시 @Embedded와 함께 사용 가능할텐데, (특정 버전의 경우 버그가 있긴 하지만요)
그렇다면 record로 어떻게 정의하는지 알아보고, 그렇게 수정해보도록 할까요?

Firedrago95 replied:
값 객체를 record로 변환해보았습니다.

[refactor: 데이터 객체를 레코드 타입으로 전환](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/c2b04caf16cf7d39be6cbc366b59e86efbb28451)

jamie9504 commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -8,30 +15,50 @@
+
+    private LocalDate date;
+
+    @ManyToOne
```
`@ManyToOne`의 패치 전략은 무엇이 있으며, 기본 전략은 무엇인가요?

Firedrago95 replied:
https://github.com/woowacourse/spring-roomescape-waiting/pull/184#discussion_r2092547089

jamie9504 replied:
> @OneToMany의 fetch 기본값은 LAZY입니다. 즉, 연관된 엔티티는 실제로 접근할 때 조회됩니다. 만약 fetch = EAGER로 설정하면, 모든 Member를 조회할 때 각 Member가 가진 Reservation도 즉시 함께 조회되며, 이로 인해 N+1 문제를 유발할 수 있습니다.
> 
> 이러한 이유로 JPA는 ~ToMany 관계에서는 기본적으로 LAZY를 사용하고, 반대로 @ManyToOne, @OneToOne 같은 ~ToOne 관계에서는 기본값이 EAGER입니다. ~ToOne 관계는 일반적으로 추가 조회가 한 건이기 때문에 성능 부담이 적어 기본적으로 즉시 로딩을 사용합니다.

마찬가지로 기본 전략을 사용한 이유는 무엇인가요?

Firedrago95 replied:
`@ManyToOne` 관계는 추가 조회가 한 건만 이루어지고, 조회 시 항상 연관 엔티티가 필요하고, 즉시 로딩이 성능상 큰 문제가 되지 않는다고 판단되어 기본 전략을 유지하였습니다.

jamie9504 replied:
1. 해당 애너테이션의 경우 N+1이 발생할 여지는 없나요?
2. EAGER 전략과 LAZY 전략은 각각 어떻게 동작하나요? 만약 LAZY로 변경한다면 어떠한 차이가 있을까요?

Firedrago95 replied:
https://github.com/woowacourse/spring-roomescape-waiting/pull/184#discussion_r2097411086

jamie9504 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -16,21 +38,14 @@ private Member(final Long id, final Name name, final Email email, final Password
         return new Member(null, new Name(name), new Email(email), new Password(password));
     }
 
+    public void addReservation(final Reservation reservation) {
```
이 메서드를 다른 개발자가 다른 곳에서 사용하면, 설계시 의도한 대로 동작할까요?
어떻게 개선해볼 수 있을까요?

Firedrago95 replied:
연관관계의 주인이 아닌 엔티티에  addReservation() 같은 메서드가 있다면, db에 반영되지 못하는 위험이 있을 수 있을 것 같습니다. 연관관계의 주인 엔티티를 기준으로 변경감지가 실행되니까요. 그러면 연관관계를 설정하는 메서드는 Reservation 같은 연관관계 주인 엔티티에서 설정해줘야 한다고 생각했습니다. 그래서 Reservation이 생성될때 Member의 reservations에 함께 할당되도록 변경하였습니다.

[refactor : 연관관계 설정을 Reservation 에서 하도록 변경](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/54dce0140b450c24b1c6ff6191d7ebbf1b1dadf6)

jamie9504 replied:
Reservation에선 `member.getReservations().add(this);`으로 set을 하고 있는데
마찬가지로 service나 다른 domain에서 해당 Member에 대고 `getReservations().add(reservation)`을 하면 제가 위에서 드렸던 질문이 동일하게 동작하는 것 아닌가요?

Firedrago95 replied:
결국 해결할 수 있는 방법을 생각해내지 못하였습니다. 양방향 연관관계를 사용하는 이상 연관관계를 일치하기 위해서 반드시 외부에서 주입해줄 수 밖에 없다는 생각이 들었습니다. 이 점이 양방향 관계의 위험성이 아닌가 생각하게 되었습니다. 아니면 제가 방법을 찾아내지 못한것일까요?

jamie9504 replied:
`연관관계 편의 메서드`에 대해 알아볼까요?

Firedrago95 replied:
처음 제가 사용한 메서드가 `연관관계 편의메서드`라고 생각했는데 지적하신대로 잘못된 외부호출을 막을 수 있는 방법을 모르겠습니다.

jamie9504 replied:
이 부분은 슬랙으로 이야기 나누고, 다음 단계 PR에 나눴던 대화를 정리하여 남겨주는 것으로 할게요.
코드에 대한 질문은 되도록 리뷰 코멘트로 받고 있는데, 그 이유는  슬랙으로 나눈 대화는 다른 크루들 or 미래에 본인이 확인할 수 없기 때문입니다.

jamie9504 reviewed (changes requested):
안녕하세요, 드라고. 리뷰어 제이미입니다.

만나서 반갑습니다.

코멘트 남겨두었습니다.

Firedrago95 committed:
docs: readme 작성

Firedrago95 committed:
refactor: Transactional 어노테이션 도입

Firedrago95 committed:
refactor : 연관관계 설정을 Reservation 에서 하도록 변경

jamie9504 commented on src/main/java/roomescape/admin/service/AdminService.java:
```diff
@@ -18,11 +19,13 @@ public boolean isExistsByEmail(final String email) {
         return adminRepository.existsByEmail(email);
     }
```
해당 메서드에는 `@Transactional` 애너테이션을 사용하지 않은 이유는 무엇인가요?

---

`@Transactional` 애너테이션에 대해 조금 더 심화 학습을 해보면 좋겠는데, 드라고가 생각한 적정 수준까지 학습해보고 정리해보세요.

Firedrago95 replied:
- 해당 메서드에 `@Transactional` 에너테이션을 추가하였습니다.

Firedrago95 replied:
## `@Transaction` 트랜잭션 전파 
  - 트랜잭션의 propagation 속성을 통해 내부 메서드 실행시 이미 트랜잭션이 실행되고 있는 경우를 설정해 줄 수 있습니다. 
  required : 기본속성으로 이미 트랜잭션이 존재하는 경우, 하나의 트랜잭션으로 처리한다. (여러 리포지토리 의존하는 경우)
  required_new : 기존 트랜잭션을 보류시키고 새로운 트랜잭션을 생성한다. (로깅은 성공하고 비지니스로직은 실패 하는 경우)
  nested : 중첩 트랜잭션을 생성한다. 자손의 롤백에도 부모 트랜잭션은 커밋이 가능하지만, 부모 트랜잭션이 롤백되는 경우 자손 트랜잭션은 커밋될 수 없다.
 
 ## 트랜잭션 내부 호출
 - 스프링은 프록시 객체를 생성하여 `@Transactional`을 실행하는데, 서비스 내부에서 다른 메서드를 호출 할 경우 프록시 객체를 통한 호출이 적용되지 않기때문에 트랜잭션이 적용되지 않습니다. 내부호출이 일어나지 않도록 클래스 분리를 고려해야 합니다.

jamie9504 replied:
1. 클래스에 `@Transactional` 메서드를 선언하면 어떠한 효과가 있나요?
2. 메서드에 각각 설정하는 것, 클래스 + 메서드를 합쳐서 설정하는 것 등 여러 방법이 있을텐데 어떤 것이 드라고의 취향인가요?

Firedrago95 replied:
클래스에 `@Transactional`을 선언하면 메서드에 공통적으로 적용된다는 사실을 배웠습니다. 대부분의 작업이 읽기전용인 경우가 많으므로 클래스에 readOnly = true 옵션을 선언하고 삽입,수정, 삭제 작업의 경우 메서드에 `@Transactional`을 선언하는 방식이 더 가독성 좋은 방식이라는 생각이 들었습니다. 
[(readOnly = true) 어노테이션 클래스 레벨로 이동](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/0412ed57b460b4b0ad48c71717e0206a48bf369f)

jamie9504 reviewed (approved):
안녕하세요, 드라고. 리뷰어 제이미입니다.

코멘트 남겨두었습니다.

jamie9504 reviewed (changes requested):
안녕하세요, 드라고. 리뷰어 제이미입니다.

코멘트 남겨두었습니다.

(Approve로 잘못 보내서 RC 다시 보냅니다.)
다음 요청 때는 Approve & Merge 예정입니다.

Firedrago95 committed:
"refactor: 이메일 중복 확인 메소드에 readOnly 추가"

jamie9504 commented on src/main/java/roomescape/admin/domain/Admin.java:
```diff
@@ -1,11 +1,22 @@
+    private String email;
+    private String password;
+
+    public Admin() {
```
해당 생성자를 꼭 public으로 열어야 할까요?

Firedrago95 replied:
JPA는 기본생성자를 통해 엔티티를 생성하고, 리플렉션을 사용하여 필드값을 초기화 해준다는 사실을 알게되었습니다. 리플렉션을 사용할때 protected 도 접근 할 수 있으므로 예상밖의 엔티티가 생성되는 것을 막기 위해 `protected` 로 기본생성자를 선언해주는 것이 좋다는 것을 배웠습니다.

[refactor: 기본 생성자의 접근 제어자를 protected로 변경](https://github.com/woowacourse/spring-roomescape-waiting/pull/184/commits/157746ec08db13c973ddd8ffc726493995f293e7)

jamie9504 replied:
새롭게 접하는 부분들에 대한 명확한 사용법을 리뷰가 없이도 익히려면 어떻게 해야할지도 고민해보세요.
항상 리뷰가 있는 상황에서 학습을 할 수 없으므로 미리미리 준비해두어야 좋습니다.

jamie9504 reviewed (changes requested):
안녕하세요, 드라고. 리뷰어 제이미입니다.

이번 단계에서 짚고 넘어가면 좋을 부분들이 남아있는 것 같아서,
한 번 더 코멘트 남겨두었습니다.

다음 번에는 Approve & Merge 진행할게요.

Firedrago95 committed:
refactor: 인기 테마 조회 로직 수정

- 인기 테마 조회 시 인기 개수(limit) 로직을 서비스 레이어로 이동
- JPQL 쿼리로 변경하여 불필요한 limit 제거
- 관련된 인터페이스 및 메서드 시그니처 수정

Firedrago95 committed:
feat: StatusConverter 추가 및 @Convert 적용

- StatusConverter 클래스 추가하여 Status와 DB 컬럼 간 변환 구현
- Reservation 엔티티에서 @Enumerated를 @Convert로 변경하여 StatusConverter 적용

Firedrago95 committed:
feat: Member 엔티티의 fetch 전략 수정

- reservations 컬렉션의 fetch 옵션을 LAZY에서 기본값으로 변경하여 코드 단순화

Firedrago95 committed:
refactor: 데이터 객체를 레코드 타입으로 전환

- Email, Name, Password 클래스를 레코드(Record)로 변경
- 불필요한 getter, equals, hashCode 메서드 제거
- 생성자 검증 메서드를 static으로 변경하여 코드 간결화

Firedrago95 committed:
refactor: 기본 생성자의 접근 제어자를 protected로 변경

- Member, Admin, Reservation, Theme, ReservationTime의 기본 생성자를 public에서 protected로 변경
- 불필요한 외부 접근을 제한하고 객체 생성의 의도를 명확히 하기 위함

Firedrago95 committed:
refactor: @Transactional(readOnly = true) 어노테이션 클래스 레벨로 이동

- 서비스 클래스에서 메서드별로 선언된 @Transactional(readOnly = true) 어노테이션을 클래스 레벨로 이동
- 코드 간소화 및 중복 제거를 통해 가독성과 유지보수성 향상
- Reservation 클래스의 멤버 필드 선언 위치 수정

jamie9504 commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -1,37 +1,64 @@
+
+    private void setMember(Member member) {
         this.member = member;
+        member.getReservations().add(this);
```
디미터의 법칙. 레벨1에서 배웠던 부분도 잊지 마세요.

jamie9504 commented on src/main/java/roomescape/reservation/presentation/MemberReservationController.java:
```diff
@@ -0,0 +1,53 @@
+        this.reservationService = reservationService;
+    }
+
+    @PostMapping("/reservations")
```
URL만 보았을 때, 어떠한 동작을 할지 충분히 유추할 수 있을까요?

jamie9504 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -43,63 +51,78 @@ public ReservationService(
+        Theme theme = themeRepository.findById(request.themeId())
+            .orElseThrow(() -> new ThemeException("테마를 찾을 수 없습니다."));
+        Member member = memberRepository.findById(memberId)
+            .orElseThrow(() -> new BusinessException("멤버를 찾을 수 없습니다."));
```
ReservationTimeException, ThemeException는 따로 만들었지만 Member에 대한 것은 BusinessException를 사용한 이유는 무엇일까요?

jamie9504 commented on src/test/java/roomescape/reservation/service/ReservationServiceTest.java:
```diff
@@ -1,83 +1,59 @@
 package roomescape.reservation.service;
 
+import static org.assertj.core.api.Assertions.assertThat;
```
JPA를 도입했는데, JPA에 대한 테스트가 하나도 존재하지 않는 이유는 무엇인가요?

jamie9504 reviewed (approved):
안녕하세요, 드라고. 리뷰어 제이미입니다.

이번 단계는 충족한 것 같아 Approve & Merge합니다.

코멘트 남겨두었습니다.

남겨둔 코멘트 중 추가적으로 이야기를 나눌 부분이 있다면 다음 PR 본문에 적어주세요.
머지한 PR은 추가적으로 확인하지 않습니다.
